File: misli/misli/basic_classes/point2d.py

from __future__ import annotations
from typing import Tuple
import math


class Point2D:
    def __init__(self, x: float = 0, y: float = 0):
        self._x = x
        self._y = y

    def __repr__(self):
        return '<Point x=%s y=%s>' % (self.x(), self.y())

    def __eq__(self, other: Point2D) -> bool:
        return self.x() == other.x() and self.y() == other.y()

    def __add__(self, other: Point2D) -> Point2D:
        return Point2D(self.x() + other.x(), self.y() + other.y())

    def __sub__(self, other: Point2D) -> Point2D:
        return Point2D(self.x() - other.x(), self.y() - other.y())

    def __truediv__(self, k) -> Point2D:
        return Point2D(self.x() / k, self.y() / k)

    def __round__(self):
        return Point2D(round(self.x()), round(self.y()))

    def __mul__(self, k):
        return Point2D(self.x() * k, self.y() * k)

    def __rmul__(self, k):
        return self.__mul__(k)

    def x(self) -> float:
        return self._x

    def y(self) -> float:
        return self._y

    def set_x(self, new_x: float):
        self._x = new_x

    def set_y(self, new_y: float):
        self._y = new_y

    def as_tuple(self) -> Tuple[float, float]:
        return (self._x, self._y)

    def distance_to(self, point: Point2D) -> float:
        distance = math.sqrt(
            (self.x() - point.x())**2 + (self.y() - point.y())**2)
        return distance

    def rotated(self, radians: float, origin: Point2D) -> Point2D:
        """Rotate the point around a given origin.
        """
        adjusted_x = (self.x() - origin.x())
        adjusted_y = (self.y() - origin.y())
        cos_rad = math.cos(radians)
        sin_rad = math.sin(radians)
        qx = origin.x() + cos_rad * adjusted_x + sin_rad * adjusted_y
        qy = origin.y() + -sin_rad * adjusted_x + cos_rad * adjusted_y

        return Point2D(qx, qy)

File: web_app/src/util/Point2D.ts
// import { Point2D } from '../types/util';
export type PointData = [number, number];

export class Point2D {
  constructor(public x: number = 0, public y: number = 0) { }

  static fromData(obj: PointData): Point2D {
    return new Point2D(obj[0], obj[1]);
  }

  data(): PointData {
    return [this.x, this.y];
  }

  copy(): Point2D {
    return new Point2D(this.x, this.y);
  }

  toString(): string {
    return `<Point x=${this.x} y=${this.y}>`;
  }

  equals(other: Point2D): boolean {
    return this.x === other.x && this.y === other.y;
  }

  add(other: Point2D): Point2D {
    return new Point2D(this.x + other.x, this.y + other.y);
  }

  subtract(other: Point2D): Point2D {
    return new Point2D(this.x - other.x, this.y - other.y);
  }

  divide(k: number): Point2D {
    return new Point2D(this.x / k, this.y / k);
  }

  round(): Point2D {
    return new Point2D(Math.round(this.x), Math.round(this.y));
  }

  multiply(k: number): Point2D {
    return new Point2D(this.x * k, this.y * k);
  }

  distanceTo(point: Point2D): number {
    const distance = Math.sqrt(
      Math.pow(this.x - point.x, 2) + Math.pow(this.y - point.y, 2)
    );
    return distance;
  }

  rotated(radians: number, origin: Point2D): Point2D {
    const adjustedX = this.x - origin.x;
    const adjustedY = this.y - origin.y;
    const cosRad = Math.cos(radians);
    const sinRad = Math.sin(radians);
    const qx = origin.x + cosRad * adjustedX + sinRad * adjustedY;
    const qy = origin.y - sinRad * adjustedX + cosRad * adjustedY;
    return new Point2D(qx, qy);
  }
}
